File: src/main/java/com/imooc/flink/java/course05/CustomParallelSourceFunction.java
package com.imooc.flink.java.course05;

import org.apache.flink.streaming.api.functions.source.ParallelSourceFunction;

/**
 * 自定义SourceFunction,支持并行
 * 每隔1s,产生一个自增的数字
 */
public class CustomParallelSourceFunction implements ParallelSourceFunction<Long> {
    Long count = 1L;
    boolean isRunning = true;
    @Override
    public void run(SourceContext<Long> ctx) throws Exception {
        while (true){
            ctx.collect(count);
            count ++;
            Thread.sleep(1000); // 睡眠1s,
        }
    }

    @Override
    public void cancel() {
        isRunning = false;
    }
}

File: src/main/scala/com/imooc/flink/scala/course05/CustomParallelSourceFunction.scala
package com.imooc.flink.scala.course05

import org.apache.flink.streaming.api.functions.source.{ParallelSourceFunction, SourceFunction}
//隐式转换
import org.apache.flink.api.scala._

/**
 * 自定义SourceFunction,支持并行
 * 每隔1s,产生一个自增的数字
 */
class CustomParallelSourceFunction extends ParallelSourceFunction[Long]{
  var count = 1L
  var isRunning = true
  override def cancel(): Unit = {
    isRunning = false
  }
  override def run(ctx: SourceFunction.SourceContext[Long]): Unit = {
    while (isRunning){
      ctx.collect(count)
      count += 1
      Thread.sleep(1000) // 睡眠1s,
    }
  }
}
